cmd/git-sync/main.go

```diff
@@ -569,66 +563,108 @@ func revIsHash(ctx context.Context, rev, gitRoot string) (bool, error) {
 	return strings.HasPrefix(output, rev), nil
 }
 
-// syncRepo syncs the branch of a given repository to the destination at the given rev.
-// returns (1) whether a change occured, (2) the new hash, and (3) an error if one happened
-func syncRepo(ctx context.Context, repo, branch, rev string, depth int, gitRoot, dest string) (bool, string, error) {
-	target := path.Join(gitRoot, dest)
-	gitRepoPath := path.Join(target, ".git")
-	var hash string
-	_, err := os.Stat(gitRepoPath)
+// syncRepo syncs the rev of a given repository to the destination directory.
+// This returns 1) whether a change occured and 2) an error if one happened.
+func syncRepo(ctx context.Context, repo, rev string, depth int, gitRoot, dest string) (bool, string, error) {
+	repoDotGitPath := path.Join(gitRoot, ".git")
+
+	var hashToSync string
+
+	// Figure out if this is the first sync or not.
+	_, err := os.Stat(repoDotGitPath)
 	switch {
 	case os.IsNotExist(err):
 		// First time. Just clone it and get the hash.
-		err = cloneRepo(ctx, repo, branch, rev, depth, gitRoot)
-		if err != nil {
-			return false, "", err
-		}
-		hash, err = localHashForRev(ctx, rev, gitRoot)
+		hash, err := cloneAndHash(ctx, repo, rev, depth, gitRoot)
 		if err != nil {
 			return false, "", err
 		}
+		hashToSync = hash
 	case err != nil:
-		return false, "", fmt.Errorf("error checking if repo exists %q: %v", gitRepoPath, err)
+		return false, "", fmt.Errorf("error checking if repo exists %q: %v", gitRoot, err)
 	default:
-		// Not the first time. Figure out if the ref has changed.
-		local, remote, err := getRevs(ctx, target, branch, rev)
-		if err != nil {
-			return false, "", err
-		}
-		if local == remote {
-			log.V(1).Info("no update required", "rev", rev, "local", local, "remote", remote)
-			return false, "", nil
+		// Make sure the directory we found is actually usable.
+		if err := sanityCheck(ctx, gitRoot, dest); err != nil {
```

Comment:
Running the sanityCheck every time feels quite expensive. We are running this on each sync run even if nothing changes aka local == remote.
I see the reasoning, but this might be problematic especially considering the normal use case of it being run in a small sidecar.
What was the specific thinking here? Validity and better error recovery > performance considerations?